File: neetcode150/medium/567-Permutation-in-String.py
#!/usr/bin/python3

class Solution:
  # Inital solution 
  def checkInclusion(self, s1: str, s2: str) -> bool:
    if len(s1) > len(s2):
      return False

    freq = {}

    for c in s1:
      freq[c] = freq.get(c, 0) + 1
    
    i = 0
    j = len(s1)

    while j <= len(s2):
      curr = {}
      for l in range(j - i):
        curr[s2[i + l]] = curr.get(s2[i + l], 0) + 1
      if freq == curr:
        return True
      i += 1
      j += 1
    return False

  # Optimal solution O(n)
  def checkInclusion(self, s1: str, s2: str) -> bool:
    if len(s1) > len(s2):
      return False
        
    s1freq, s2freq = [0] * 26, [0] * 26
        
    for i in range(len(s1)):
      s1freq[ord(s1[i]) - ord('a')] += 1
      s2freq[ord(s2[i]) - ord('a')] += 1
            
    match = 0
    for i in range(26):
      if s1freq[i] == s2freq[i]:
        match += 1
                
    l = 0
    for r in range(len(s1), len(s2)):
      if match == 26:
        return True
            
      index = ord(s2[r]) - ord('a')
      s2freq[index] += 1
      if s1freq[index] == s2freq[index]:
        match += 1
      elif s1freq[index] + 1 == s2freq[index]:
        match -= 1
                
      index = ord(s2[l]) - ord('a')
      s2freq[index] -= 1
      if s1freq[index] == s2freq[index]:
        match += 1
      elif s1freq[index] - 1 == s2freq[index]:
        match -= 1
      l += 1
            
    return match == 26
  

File: neetcode150/medium/567-Permutation-in-String.cpp
#include <vector>
#include <string>

using namespace std;

/**
 * \c Solution1
 * 
 * Sorting Approach:
 * Sort input string and compare it to the sorted substr in s2
 * 
 * Time  : O(M * N log N)
 * Space : O(N)
 */
class Solution1 {
public:
    bool checkInclusion(string s1, string s2) {
        sort(s1.begin(), s1.end());

        int i = 0;
        int n = s1.size();
        while ((i+n-1) < s2.size()) {
            string substr = s2.substr(i, s1.size());
            sort(substr.begin(), substr.end());
            if (substr == s1) {
                return true;
            }
            ++i;
        }
        return false;
    }
};

/**
 * \c Solution2
 * 
 * Optimal Approach
 * 1. Store the character frequency of s1 and s2.substr(s1.size())
 * 2. Use cool increment/decrement check approach 
 * 
 * Time  : O(N)
 * Space : O(1)
 */
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        if (s1.size() > s2.size()) return false;

        vector<int> s1Freq(26, 0);
        vector<int> s2Freq(26, 0);
        for (int i = 0; i < s1.size(); ++i) {
            ++s1Freq[s1[i]-'a'];
            ++s2Freq[s2[i]-'a'];
        }

        if (s1Freq == s2Freq) return true;

        for (int j = s1.size(); j < s2.size(); ++j) {
            ++s2Freq[s2[j]-'a'];
            --s2Freq[s2[j-s1.size()]-'a'];

            if (s1Freq == s2Freq) return true;
        }
        return false;
    }
};
